Enablement for using states to manage AWS EC2 instances

tkwilliams opened the pull request:
Trying to entirely replace our CloudFormation VPC buildout mess with 'salt-call state.sls aws.vpc' and I'm getting REALLLLY close now...
This changeset adds the following:
- Add optional ability to detect instance names which are in 'terminated'-type states and thus safe to re-use
- Fix some stacktraces from un-checked get_id calls
- Add vpc_id/vpc_name options to various things that need them
- Add availability_zone param to subnets so they can be attached to VPCs
- boto_ec2.instance_present()
- boto_ec2.instance_absent()
- Various security_group fixes
- Update boto_ec2.run() to return the new instance_id rather than just True
- Add almost all available boto.ec2.run_instances() params to boto_ec2.run()
- Add support for passing 'subnet_name' as an alternative to 'subnet_id' for boto_ec2.run()
- Fix stacktrace when calling boto_ec2.run()
- Allow 'instance_name' (vs. 'instance_id') in route option to boto_vpc.route_table_present()
- Add dhcp_options manipulation functions to the module, so that we can then do dhcp_options_present() in the boto_vpc state
- and WTH, some spelling and capitalization fixes...

tkwilliams committed:
Add dhcp_options manipulation functions to the module, so that we can then do dhcp_options_present() in the boto_vpc state

tkwilliams committed:
Allow 'instance_name' (vs. 'instance_id') in route option to boto_vpc.route_table_present()

tkwilliams committed:
Fix stacktrace when calling boto_ec2.run()

tkwilliams committed:
Add support for passing 'subnet_name' as an alternative to 'subnet_id' for boto_ec2.run()

tkwilliams committed:
Add almost all available boto.ec2.run_instances() params to boto_ec2.run()

tkwilliams committed:
Update boto_ec2.run() to return the new instance_id rather than just True

tkwilliams committed:
Merge remote-tracking branch 'upstream/develop' into d.boto_vpc.dhcp_options_stuff

tkwilliams committed:
Enablement for using states to manage AWS EC2 instances
- ability to detect instance_names which are in 'terminated'-type states and thus safe to re-use
- fix some stacktraces from un-checked get_id calls
- add vpc_id/vpc_name options to various things that need them
- add availability_zone param to subnets so they can be attached to VPCs
- boto_ec2.instance_present()
- boto_ec2.instance_absent()
- various security_group fixes
- and WTH, some spelling and capitalization fixes...

tkwilliams commented:
@ryan-lane  This is the biggest thing I've done so far - I'd be glad for any feedback or criticism (or what the heck, advice) you might have about my goals :)

tkwilliams committed:
Resolve pylint whinging :)  I strongly disagree with several PEP "standards" but... what can you do?

ryan-lane commented on salt/modules/boto_secgroup.py:
```diff
@@ -255,8 +273,7 @@ def convert_to_group_ids(groups, vpc_id=None, region=None, key=None, keyid=None,
                       ' group name {0}'.format(group))
-            group_id = get_group_id(group, vpc_id, region, key, keyid,
-                                    profile, vpc_name)
+            group_id = get_group_id(group, vpc_id, vpc_name, region, key, keyid, profile)
```
Since you've been shifting around args, can you make sure all of these are named, like group=group, vpc_id=vpc_id, etc? It'll make it more reliable in the future and easier to review.

ryan-lane replied:
Same note for all other locations where you're calling a function where you've moved around the args.

tkwilliams replied:
Oook, I thought I had done all of those - thanks for catching :)

t.

On Tue, Oct 6, 2015 at 6:35 PM, Ryan Lane <notifications@github.com> wrote:

> In salt/modules/boto_secgroup.py
> <https://github.com/saltstack/salt/pull/27728#discussion_r41331623>:
>
> > @@ -255,8 +273,7 @@ def convert_to_group_ids(groups, vpc_id=None, region=None, key=None, keyid=None,
> >          else:
> >              log.debug('calling boto_secgroup.get_group_id for'
> >                        ' group name {0}'.format(group))
> > -            group_id = get_group_id(group, vpc_id, region, key, keyid,
> > -                                    profile, vpc_name)
> > +            group_id = get_group_id(group, vpc_id, vpc_name, region, key, keyid, profile)
>
> Same note for all other locations where you're calling a function where
> you've moved around the args.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/27728/files#r41331623>.
>

tkwilliams replied:
I had actually added the args in a branch WITHOUT them (e.g. no vpc_name)
but by the time I went to merge they'd been added independently in
different locations...  I ended up having to move stuff around to merge and
likely missed some cases.

Thanks again


On Tue, Oct 6, 2015 at 6:36 PM, Thomas Williams <bif.bam@gmail.com> wrote:

> Oook, I thought I had done all of those - thanks for catching :)
>
> t.
>
> On Tue, Oct 6, 2015 at 6:35 PM, Ryan Lane <notifications@github.com>
> wrote:
>
>> In salt/modules/boto_secgroup.py
>> <https://github.com/saltstack/salt/pull/27728#discussion_r41331623>:
>>
>> > @@ -255,8 +273,7 @@ def convert_to_group_ids(groups, vpc_id=None, region=None, key=None, keyid=None,
>> >          else:
>> >              log.debug('calling boto_secgroup.get_group_id for'
>> >                        ' group name {0}'.format(group))
>> > -            group_id = get_group_id(group, vpc_id, region, key, keyid,
>> > -                                    profile, vpc_name)
>> > +            group_id = get_group_id(group, vpc_id, vpc_name, region, key, keyid, profile)
>>
>> Same note for all other locations where you're calling a function where
>> you've moved around the args.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/saltstack/salt/pull/27728/files#r41331623>.
>>
>
>

ryan-lane commented on salt/states/boto_ec2.py:
```diff
@@ -452,3 +452,151 @@ def snapshot_created(name, ami_name, instance_name, wait_until_available=True, w
+            _create = True
+
+    if _create:
+        r = __salt__['boto_ec2.run'](image_id, instance_name if instance_name else name,
```
Missing test mode check. This will create instances when people call the state with test=true

ryan-lane commented on salt/states/boto_ec2.py:
```diff
@@ -452,3 +452,151 @@ def snapshot_created(name, ami_name, instance_name, wait_until_available=True, w
+            continue
+        else:
+            try:
+                r = __salt__['boto_ec2.set_attribute'](attribute=k, attribute_value=v,
```
Also missing test mode.

ryan-lane commented on salt/states/boto_ec2.py:
```diff
@@ -452,3 +452,151 @@ def snapshot_created(name, ami_name, instance_name, wait_until_available=True, w
+        ret['comment'] = 'Termination of instance {0} via the API is disabled.'.format(instance_id)
+        return ret
+
+    r = __salt__['boto_ec2.terminate'](instance_id=instance_id, name=instance_name, region=region,
```
Also missing test mode.

tkwilliams replied:
OUCH, you're right.  Fixing now...  I've been "testing" all this time by
building and deleting a VPC repeatedly, and wasn't thinking about test mode
at all...

On Tue, Oct 6, 2015 at 6:38 PM, Ryan Lane <notifications@github.com> wrote:

> In salt/states/boto_ec2.py
> <https://github.com/saltstack/salt/pull/27728#discussion_r41331874>:
>
> > +    exists = __salt__['boto_ec2.exists'](instance_id=instance_id, region=region,
> > +                                         key=key, keyid=keyid, profile=profile)
> > +    if not exists:
> > +        ret['result'] = True
> > +        ret['comment'] = 'Instance {0} is already gone.'.format(instance_id)
> > +        return ret
> > +
> > +    ### Honor 'disableApiTermination' - if you want to override it, first use set_attribute() to turn it off
> > +    no_can_do = __salt__['boto_ec2.get_attribute']('disableApiTermination', instance_id=instance_id,
> > +                                                   region=region, key=key, keyid=keyid, profile=profile)
> > +    if no_can_do.get('disableApiTermination') == True:
> > +        ret['result'] = False
> > +        ret['comment'] = 'Termination of instance {0} via the API is disabled.'.format(instance_id)
> > +        return ret
> > +
> > +    r = __salt__['boto_ec2.terminate'](instance_id=instance_id, name=instance_name, region=region,
>
> Also missing test mode.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/27728/files#r41331874>.
>

ryan-lane commented on salt/states/boto_vpc.py:
```diff
@@ -230,6 +230,190 @@ def absent(name, tags=None, region=None, key=None, keyid=None, profile=None):
+
+    # boto provides no "update_dhcp_options()" functionality, and you can't delete it if
+    # it's attached, and you can't detach it if it's the only one, so just check if it's
+    # there or not, and make no effort to validate it's actual settings... :(
```
Can you add more than one set of dhcp options? One approach is to create and attach new options, then to detach and delete the old ones. We do this with launch configs in boto_asg.

ryan-lane replied:
This approach is also fine, but you should document this behavior so that end-users know this is what it does.

ryan-lane commented:
Added some comments, but this is looking really good so far!

tkwilliams commented:
Oh, heh, the comment about PEP8 was WRT it moaning about the following:
```
foo = { 'xoxoxxo'       : bar1,
        'reallylong'    : bar2,
        'a'             : bar3,
        'extrasuperlong': bar4
      }
```
Which is MUCH more legible than the PEP-ified version.  They seem to assume that everyone on earth is using a color-highlighting IDE where white-space possibly isn't quite as important a signifier, but I use uncolored vim; and sorry but many PEP requirements make for VERY VERY extra-hard-to-read monochrome code... :)

</complaintMode>

tkwilliams committed:
- Add test=True cases for new functions
- Add pass-through args for modified functions
- Describe limitations of current dhcp_options code
- versionadded:: Boron tags for new funcs
- Some formatting cleanup

tkwilliams committed:
And then fix the bits I broke fixing the OTHER stuff...

tkwilliams committed:
Argh!  One.  Final.  Bug.  I hope...

tkwilliams committed:
...and one last PEPification

rallytime commented on salt/modules/boto_vpc.py:
```diff
@@ -1227,6 +1227,43 @@ def create_dhcp_options(domain_name=None, domain_name_servers=None, ntp_servers=
+                     region=None, key=None, keyid=None, profile=None):
+    '''
+    Return a dict with the current values of the requested DHCP options set
+
```
Can you add a `.. versionadded:: Boron` tag here?

tkwilliams replied:
You betcha.  Done.

rallytime commented on salt/states/boto_vpc.py:
```diff
@@ -230,6 +230,195 @@ def absent(name, tags=None, region=None, key=None, keyid=None, profile=None):
+    Note that the current implementation only SETS values during option set
+    creation.  It is unable to update option sets in place, and thus merely
+    verifies the set exists via the given name and/or dhcp_options_id param.
+
```
`.. versionadded:: Boron` here too.

rallytime commented on salt/states/boto_vpc.py:
```diff
@@ -230,6 +230,195 @@ def absent(name, tags=None, region=None, key=None, keyid=None, profile=None):
+def dhcp_options_absent(name=None, dhcp_options_id=None, region=None, key=None, keyid=None, profile=None):
+    '''
+    Ensure a set of DHCP options with the given settings exist.
+
```
`.. versionadded:: Boron` here as well.

rallytime commented:
@tkwilliams Thanks for all of this awesome work! Outstanding!

I only had a couple of small change suggestions. I think there is one more comment from @ryan-lane above that may need to be addressed. Otherwise, if he's good with everything, this is looking good to go in.

tkwilliams committed:
Missed a ".. versionadded:: Boron" tag in salt/modules/boto_vpc.py

tkwilliams committed:
...and a couple more versionadded tags.

tkwilliams commented:
Hi Nicole, thanks for the kind words.

I /think/ I've addressed all of @ryan-lane's concerns above -- the one
outstanding I haven't fixed is his suggestion for multiple DHCP options.
Following his second comment on it, I've updated the doc-strings to mention
the current, "limited" behaviour, and also added a TODO comment to remind
me to get to it soon.

I'm pretty comfortable with the current code, and since the DHCP funcs are
new features anyway, I'm hoping that "limited" is better than none (in the
short term) pending another chunk of free time on my part :)

@ryan-lane - that said, if I did miss one of your suggestions, please speak
up :)

Thanks again!

t.



On Wed, Oct 7, 2015 at 11:22 AM, Nicole Thomas <notifications@github.com>
wrote:

> @tkwilliams <https://github.com/tkwilliams> Thanks for all of this
> awesome work! Outstanding!
>
> I only had a couple of small change suggestions. I think there is one more
> comment from @ryan-lane <https://github.com/ryan-lane> above that may
> need to be addressed. Otherwise, if he's good with everything, this is
> looking good to go in.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/27728#issuecomment-146228939>.
>

ryan-lane commented on salt/states/boto_ec2.py:
```diff
@@ -452,3 +452,176 @@ def snapshot_created(name, ami_name, instance_name, wait_until_available=True, w
+                     attributes=None, target_state=None, region=None, key=None,
+                     keyid=None, profile=None):
+    ### TODO - implement 'target_state={running, stopped}'
+    ### TODO - Describe all the above params :-(
```
Would definitely be nice to have some basic arg docs before this hits a released version.

tkwilliams replied:
Added :)

ryan-lane commented:
Looks great!

tkwilliams committed:
Parameter defs for new functions per @ryan-lane's suggestion.

tkwilliams committed:
Merge branch 'develop' into d.boto_vpc.dhcp_options_stuff

Conflicts:
	salt/modules/boto_secgroup.py
	salt/states/boto_secgroup.py

jfindlay commented:
#27729

twellspring commented:
#29408 provides documentation update for the above change 
  'subnet_name' as an alternative to 'subnet_id'
